File: src/apple-tv/device.ts
import { AIRPLAY_SERVICE, AppleTV, COMPANION_LINK_SERVICE, ConnectionRecovery, type DiscoveryResult, type MdnsService, mdnsUnicast, Proto } from '@basmilius/apple-sdk';
import { DiscoverableDevice } from '../base';
import { AirPlayLogic } from '../logic';
import { capabilityToRepeatMode, getAccessoryCredentialsFromDevice } from '../utils';
import type AppleTVDriver from './driver';
import type Homey from 'homey';

const RECONNECT_INTERVAL = 15 * 60 * 1000;
const SLOW_RECOVERY_INTERVAL = 2 * 60 * 1000;
const SLOW_RECOVERY_MAX_ATTEMPTS = 15;

const CAPABILITIES = [
    'speaker_album',
    'speaker_artist',
    'speaker_duration',
    'speaker_next',
    'speaker_playing',
    'speaker_position',
    'speaker_prev',
    'speaker_track',
    'speaker_repeat',
    'speaker_shuffle',
    'artwork_url',
    'artwork_url_cloud',
    'artwork_url_local',
    'onoff',
    'power',
    'volume_down',
    'volume_mute',
    'volume_up',
    'remote_up',
    'remote_down',
    'remote_left',
    'remote_right',
    'remote_select',
    'remote_home',
    'remote_back',
    'remote_playpause',
    'now_playing_app',
    'button.restart'
];

export default class AppleTVDevice extends DiscoverableDevice<AppleTVDriver> {
    get airplayLogic(): AirPlayLogic {
        return this.#airplayLogic;
    }

    get discoveryResultAirPlay(): DiscoveryResult {
        return this.discoveryResults[AIRPLAY_SERVICE];
    }

    get sdk(): AppleTV {
        if (!this.#tv) {
            throw new Error('Apple TV SDK device is not initialized.');
        }

        return this.#tv;
    }

    get services(): Record<string, Homey.DiscoveryStrategy> {
        return this.#services;
    }

    #airplayLogic!: AirPlayLogic;
    #airplayRecovery?: ConnectionRecovery;
    #companionLinkFailed = false;
    #companionLinkRecovery?: ConnectionRecovery;
    #companionLinkRetried = false;
    #connectedOnce = false;
    #slowRecoveryAttempt = 0;
    #slowRecoveryTimer: ReturnType<typeof setTimeout> | null = null;
    #services!: Record<string, Homey.DiscoveryStrategy>;
    #tv?: AppleTV;

    async onInit(): Promise<void> {
        await this.setUnavailable('Connecting...');

        this.#services = {
            [AIRPLAY_SERVICE]: this.discovery.getStrategy('airplay')
        };

        this.#airplayLogic = new AirPlayLogic(this);
        await this.#airplayLogic.initialize();

        await this.syncCapabilities(CAPABILITIES);
        this.#registerCapabilities();
        this.#registerMaintenance();

        await super.onInit();

        this.log('Initialized.');
    }

    async onUninit(): Promise<void> {
        this.#stopSlowRecovery();
        this.#airplayRecovery?.dispose();
        this.#companionLinkRecovery?.dispose();
        await this.#airplayLogic.uninitialize();
        this.#tv?.disconnect();

        this.log('Uninitialized.');
    }

    async #connect(): Promise<void> {
        try {
            const credentials = getAccessoryCredentialsFromDevice(this);

            if (!credentials) {
                await this.setUnavailable('Cannot find credentials, please re-pair the device.');
                return;
            }

            if (!this.discoveryResultAirPlay) {
                await this.setUnavailable('Service discovery not complete, waiting for device...');
                return;
            }

            // Discover Companion Link via unicast using AirPlay's address.
            const companionLink = await this.#discoverCompanionLink(this.discoveryResultAirPlay.address);

            // Create or reconfigure the SDK device.
            if (!this.#tv) {
                this.#tv = new AppleTV({
                    airplay: this.discoveryResultAirPlay,
                    companionLink: companionLink ?? undefined
                });

                this.#airplayLogic.setDevice(this.#tv);
                this.#wireEvents();
                this.#setupRecovery(credentials);
            } else {
                this.#tv.discoveryResult = this.discoveryResultAirPlay;

                if (this.#tv.companionLink && companionLink) {
                    this.#tv.companionLink.discoveryResult = companionLink;
                }
            }

            this.log('Connecting to Apple TV...');
            await this.#tv.connect(credentials);
        } catch (err) {
            this.error('Error received', err);
            await this.setUnavailable('Cannot connect to Apple TV.');
        }
    }

    async #discoverCompanionLink(address: string): Promise<DiscoveryResult | null> {
        this.log(`Discovering Companion Link via unicast to ${address}...`);

        try {
            const results = await mdnsUnicast([address], [COMPANION_LINK_SERVICE], 5);
            const match = results.find((s: MdnsService) => s.address === address);

            if (!match) {
                this.log('Companion Link not found via unicast.');
                return null;
            }

            this.log(`Found Companion Link at ${match.address}:${match.port}.`);

            const txt = match.properties;
            const hostname = match.name.replace(/\s+/g, '-');

            return {
                id: `${hostname}.local`,
                fqdn: `${hostname}.local`,
                address: match.address,
                modelName: txt?.model ?? '',
                familyName: null,
                txt,
                service: {
                    port: match.port,
                    protocol: 'tcp',
                    type: COMPANION_LINK_SERVICE
                },
                packet: null
            } as unknown as DiscoveryResult;
        } catch (err) {
            this.error('Companion Link unicast discovery failed:', err);
            return null;
        }
    }

    #wireEvents(): void {
        if (!this.#tv) {
            return;
        }

        this.#tv.on('connected', async () => {
            this.#airplayRecovery?.reset();
            this.log('Connected to Apple TV (AirPlay).');

            if (this.#tv!.companionLink?.isConnected) {
                await this.setAvailable();
            }
        });

        this.#tv.on('disconnected', async (unexpected: boolean) => {
            if (!unexpected) {
                return;
            }

            this.log('Disconnected from Apple TV (AirPlay), reconnecting...');
            await this.setUnavailable('Disconnected from Apple TV (AirPlay), reconnecting...');
            this.#airplayRecovery?.handleDisconnect(unexpected);
        });

        this.#tv.on('power', async (state) => {
            this.log('Power state changed:', state);

            const isOn = state === 'awake' || state === 'screensaver';

            try {
                await this.setCapabilityValue('onoff', isOn);
                await this.setCapabilityValue('power', this.homey.__(isOn ? 'capability.power.on' : 'capability.power.off'));
            } catch (err) {
                this.error('Failed to set power state.', err);
            }

            if (isOn) {
                this.#airplayLogic.emitUpdate();
                return;
            }

            await this.#airplayLogic.clearNowPlaying();
        });

        if (this.#tv.companionLink) {
            this.#tv.companionLink.on('connected', async () => {
                this.#companionLinkRecovery?.reset();
                this.#companionLinkRetried = false;
                this.#stopSlowRecovery();
                this.log('Connected to Apple TV (Companion Link).');

                if (this.#tv!.airplay.isConnected) {
                    await this.setAvailable();
                }
            });

            this.#tv.companionLink.on('disconnected', async (unexpected: boolean) => {
                if (!unexpected) {
                    return;
                }

                this.log('Disconnected from Apple TV (Companion Link), reconnecting...');
                await this.setUnavailable('Disconnected from Apple TV (Companion Link), reconnecting...');
                this.#companionLinkRecovery?.handleDisconnect(unexpected);
            });
        }
    }

    #setupRecovery(credentials: NonNullable<ReturnType<typeof getAccessoryCredentialsFromDevice>>): void {
        this.#airplayRecovery?.dispose();
        this.#companionLinkRecovery?.dispose();

        this.#airplayRecovery = new ConnectionRecovery({
            maxAttempts: 3,
            baseDelay: 1000,
            reconnectInterval: RECONNECT_INTERVAL,
            onReconnect: async () => {
                const tv = this.#tv;
                if (!tv) return;
                tv.airplay.disconnectSafely();
                await this.findService(AIRPLAY_SERVICE);
                tv.discoveryResult = this.discoveryResults[AIRPLAY_SERVICE];
                tv.airplay.setCredentials(credentials);
                await tv.airplay.connect();
            }
        });

        this.#airplayRecovery.on('recovering', (attempt) => {
            this.log(`AirPlay recovery attempt ${attempt}...`);
        });

        this.#airplayRecovery.on('failed', () => {
            this.error('AirPlay recovery failed after max attempts.');
        });

        this.#companionLinkRecovery = new ConnectionRecovery({
            maxAttempts: 3,
            baseDelay: 1000,
            reconnectInterval: RECONNECT_INTERVAL,
            onReconnect: async () => {
                const tv = this.#tv;
                if (!tv?.companionLink) return;
                await tv.companionLink.disconnectSafely();
                const cl = await this.#discoverCompanionLink(this.discoveryResultAirPlay.address);

                if (!cl) {
                    throw new Error('Companion Link not found via unicast.');
                }

                tv.companionLink.discoveryResult = cl;
                await tv.companionLink.setCredentials(credentials);
                await tv.companionLink.connect();
            }
        });

        this.#companionLinkRecovery.on('recovering', (attempt) => {
            this.log(`Companion Link recovery attempt ${attempt}...`);
        });

        this.#companionLinkRecovery.on('failed', async () => {
            this.error('Companion Link recovery failed after max attempts.');
            await this.#onCompanionLinkFailed();
        });
    }

    async #startSlowRecovery(): Promise<void> {
        this.log(`Starting slow recovery phase, retrying every ${SLOW_RECOVERY_INTERVAL / 1000}s for up to ${SLOW_RECOVERY_MAX_ATTEMPTS} attempts...`);
        this.#slowRecoveryAttempt = 0;
        this.#scheduleSlowRecoveryAttempt();
        await this.setUnavailable('Device offline, retrying connection...');
    }

    #scheduleSlowRecoveryAttempt(): void {
        this.#slowRecoveryTimer = setTimeout(async () => {
            this.#slowRecoveryTimer = null;
            this.#slowRecoveryAttempt++;

            this.log(`Slow recovery attempt ${this.#slowRecoveryAttempt}/${SLOW_RECOVERY_MAX_ATTEMPTS}...`);

            try {
                const cl = await this.#discoverCompanionLink(this.discoveryResultAirPlay.address);

                if (!cl) {
                    throw new Error('Companion Link not found via unicast.');
                }

                const tv = this.#tv;

                if (!tv?.companionLink) {
                    return;
                }

                this.log(`Re-discovered Companion Link at ${cl.address}:${cl.service.port}, reconnecting...`);
                this.#companionLinkRecovery?.reset();
                tv.companionLink.discoveryResult = cl;

                const credentials = getAccessoryCredentialsFromDevice(this);

                if (credentials) {
                    await tv.companionLink.setCredentials(credentials);
                    await tv.companionLink.connect();
                }
            } catch {
                this.log(`Slow recovery attempt ${this.#slowRecoveryAttempt} failed.`);
            }

            if (this.#slowRecoveryAttempt >= SLOW_RECOVERY_MAX_ATTEMPTS) {
                this.#companionLinkFailed = true;
                this.log('Companion Link failed permanently after extended recovery period. Please restart the app.');
                await this.setUnavailable('Failed to connect to Apple TV using Companion Link. Please restart the app.');
                await this.app.appleTvFlow.triggerCompanionLinkFailed(this);
                return;
            }

            if (!this.#tv?.companionLink?.isConnected && this.#tv) {
                this.#scheduleSlowRecoveryAttempt();
            }
        }, SLOW_RECOVERY_INTERVAL);
    }

    #stopSlowRecovery(): void {
        if (this.#slowRecoveryTimer) {
            clearTimeout(this.#slowRecoveryTimer);
            this.#slowRecoveryTimer = null;
        }
        this.#slowRecoveryAttempt = 0;
    }

    async #onCompanionLinkFailed(): Promise<void> {
        if (this.#companionLinkFailed) {
            return;
        }

        if (this.#slowRecoveryTimer) {
            return;
        }

        if (!this.#companionLinkRetried) {
            this.#companionLinkRetried = true;

            this.log('Companion Link failed, attempting re-discovery before giving up...');
            await this.setUnavailable('Reconnecting to Apple TV...');

            try {
                const cl = await this.#discoverCompanionLink(this.discoveryResultAirPlay.address);

                if (!cl) {
                    throw new Error('Companion Link not found via unicast.');
                }

                const tv = this.#tv;

                if (!tv?.companionLink) {
                    return;
                }

                this.log(`Re-discovered Companion Link at ${cl.address}:${cl.service.port}, reconnecting...`);
                this.#companionLinkRecovery?.reset();
                tv.companionLink.discoveryResult = cl;

                const credentials = getAccessoryCredentialsFromDevice(this);

                if (credentials) {
                    await tv.companionLink.setCredentials(credentials);
                    await tv.companionLink.connect();
                }

                return;
            } catch {
                this.log('Re-discovery of Companion Link service failed.');
            }
        }

        await this.#startSlowRecovery();
    }

    #registerCapabilities(): void {
        this.#registerOnOff();
        this.#registerRemote();

        this.registerCapabilityListener('speaker_next', async () => {
            await this.sdk.playback.next();
        });

        this.registerCapabilityListener('speaker_prev', async () => {
            await this.sdk.playback.previous();
        });

        this.registerCapabilityListener('speaker_stop', async () => {
            await this.sdk.playback.stop();
        });

        this.registerCapabilityListener('speaker_playing', async (play: boolean) => {
            if (play) {
                await this.sdk.playback.play();
            } else {
                await this.sdk.playback.pause();
            }
        });

        this.registerCapabilityListener('volume_up', async () => {
            await this.sdk.volume.up();
        });

        this.registerCapabilityListener('volume_down', async () => {
            await this.sdk.volume.down();
        });

        this.registerCapabilityListener('volume_mute', async () => {
            await this.sdk.remote.mute();
        });

        this.registerCapabilityListener('speaker_repeat', async (value: string) => {
            await this.sdk.playback.setRepeatMode(capabilityToRepeatMode[value] ?? Proto.RepeatMode_Enum.Off);
        });

        this.registerCapabilityListener('speaker_shuffle', async (value: boolean) => {
            const mode = value ? Proto.ShuffleMode_Enum.Songs : Proto.ShuffleMode_Enum.Off;
            await this.sdk.playback.setShuffleMode(mode);
        });
    }

    #registerMaintenance(): void {
        this.registerCapabilityListener('button.restart', async () => {
            try {
                this.#stopSlowRecovery();
                this.#airplayRecovery?.dispose();
                this.#companionLinkRecovery?.dispose();
                this.#companionLinkFailed = false;
                this.#companionLinkRetried = false;
                this.#tv?.disconnect();
                this.#tv = undefined;
                await this.#airplayLogic.clearNowPlaying();
                await this.#connect();
            } catch (err) {
                this.error(err);
            }
        });
    }

    #registerOnOff(): void {
        this.registerCapabilityListener('onoff', async (value: boolean) => {
            if (value) {
                await this.sdk.power?.on();
            } else {
                await this.sdk.power?.off();
            }
        });
    }

    #registerRemote(): void {
        const keys = CAPABILITIES.filter(k => k.startsWith('remote_'));

        this.registerMultipleCapabilityListener(keys, async values => {
            values.remote_up === true && await this.sdk.remote.up();
            values.remote_down === true && await this.sdk.remote.down();
            values.remote_left === true && await this.sdk.remote.left();
            values.remote_right === true && await this.sdk.remote.right();
            values.remote_select === true && await this.sdk.remote.select();
            values.remote_home === true && await this.sdk.remote.home();
            values.remote_back === true && await this.sdk.remote.menu();
            values.remote_playpause === true && await this.sdk.remote.playPause();
        }, 0);
    }

    async onServiceFound(service: string, discoveryResult: DiscoveryResult): Promise<void> {
        await super.onServiceFound(service, discoveryResult);

        if (this.#connectedOnce) {
            return;
        }

        if (!this.discoveryResultAirPlay) {
            return;
        }

        this.#connectedOnce = true;
        await this.#connect();
    }
}

File: drivers/apple-tv/device.py
"""Apple TV device — AirPlay + Companion Link via pyatv."""

from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING

import pyatv
from pyatv.const import Protocol

from app.lib.airplay_logic import AirPlayLogic
from app.lib.discoverable_device import DiscoverableDevice

if TYPE_CHECKING:
    from pyatv.interface import AppleTV, BaseConfig

logger = logging.getLogger(__name__)

AIRPLAY_SERVICE = 'airplay'
COMPANION_SERVICE = 'companion-link'
RECONNECT_DELAY = 1.0  # seconds
# Scheduled Companion Link reconnect every 5 minutes
COMPANION_RECONNECT_INTERVAL = 5 * 60  # seconds

CAPABILITIES = [
    'speaker_album',
    'speaker_artist',
    'speaker_duration',
    'speaker_next',
    'speaker_playing',
    'speaker_position',
    'speaker_prev',
    'speaker_track',
    'artwork_url',
    'onoff',
    'power',
    'volume_down',
    'volume_mute',
    'volume_up',
    'remote_up',
    'remote_down',
    'remote_left',
    'remote_right',
    'remote_select',
    'remote_home',
    'remote_back',
    'remote_playpause',
    'now_playing_app',
    'media_type',
    'volume_set',
    'button.restart',
]


class AppleTVDevice(DiscoverableDevice):
    """Represents a single Apple TV on the local network."""

    def __init__(self, **kwargs) -> None:
        super().__init__(**kwargs)
        self._atv: AppleTV | None = None
        self._airplay_logic: AirPlayLogic | None = None
        self._is_reconnecting = False
        self._companion_reconnect_task: asyncio.Task | None = None

    @property
    def services(self) -> list[str]:
        return [AIRPLAY_SERVICE, COMPANION_SERVICE]

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def on_init(self) -> None:
        await self.set_unavailable('Connecting...')

        self._airplay_logic = AirPlayLogic(self)

        for cap in list(self.get_capabilities()):
            if cap not in CAPABILITIES:
                try:
                    await self.remove_capability(cap)
                except Exception:
                    pass
        self._register_capabilities()

        await super().on_init()
        self.log('Initialized.')

    async def on_uninit(self) -> None:
        self._stop_companion_reconnect()
        if self._airplay_logic:
            self._airplay_logic.stop()
        await self._disconnect()
        self.log('Uninitialized.')

    # ------------------------------------------------------------------
    # Discovery callback → connect
    # ------------------------------------------------------------------

    async def _on_device_found(self, config: BaseConfig) -> None:
        """Called by DiscoverableDevice when the pyatv config is ready."""
        if self._atv is not None:
            return  # Already connected
        await self._connect()

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    async def _connect(self) -> None:
        """Connect to the Apple TV using stored discovery results.

        Raises on failure so callers (SDK discovery callback or _reconnect)
        can handle it appropriately.
        """
        config = self._discovery_results.get(AIRPLAY_SERVICE)
        if config is None:
            raise RuntimeError('Cannot find AirPlay service.')

        airplay_creds = self._get_credentials('credentials')
        if airplay_creds:
            service = config.get_service(Protocol.AirPlay)
            if service:
                service.credentials = airplay_creds

        companion_creds = self._get_credentials('companion_credentials')
        if companion_creds:
            service = config.get_service(Protocol.Companion)
            if service:
                service.credentials = companion_creds

        loop = asyncio.get_running_loop()
        self._atv = await pyatv.connect(config, loop)
        self._atv.listener = self
        self._airplay_logic.set_protocol(self._atv)

        self._start_companion_reconnect()

        self.log('Connected to Apple TV.')

    async def _disconnect(self) -> None:
        self._stop_companion_reconnect()
        if self._airplay_logic:
            self._airplay_logic.stop()
        if self._atv is not None:
            try:
                self._atv.close()
            except Exception:
                pass
            self._atv = None

    # ------------------------------------------------------------------
    # pyatv DeviceListener callbacks
    # ------------------------------------------------------------------

    def connection_lost(self, exception: Exception | None) -> None:
        self.log(f'Connection lost: {exception}')
        asyncio.create_task(self._on_disconnected(unexpected=True))

    def connection_closed(self) -> None:
        self.log('Connection closed.')
        asyncio.create_task(self._on_disconnected(unexpected=False))

    async def _on_disconnected(self, unexpected: bool) -> None:
        if not unexpected or self._is_reconnecting:
            return

        self._is_reconnecting = True
        try:
            self.log('Disconnected from Apple TV, reconnecting...')
            await self.set_unavailable('Disconnected from Apple TV, reconnecting...')
            await asyncio.sleep(RECONNECT_DELAY)
            await self._disconnect()
            await self._reconnect()
        except Exception as err:
            self.error(f'Reconnection failed: {err}')
            await self.set_unavailable(f'Cannot reconnect: {err}')
        finally:
            self._is_reconnecting = False

    # ------------------------------------------------------------------
    # Scheduled Companion reconnect (every 5 min)
    # ------------------------------------------------------------------

    def _start_companion_reconnect(self) -> None:
        if self._companion_reconnect_task and asyncio.current_task() is self._companion_reconnect_task:
            return

        self._stop_companion_reconnect()

        async def _reconnect_loop() -> None:
            while True:
                await asyncio.sleep(COMPANION_RECONNECT_INTERVAL)
                self.log('Scheduled Companion Link reconnect...')
                try:
                    await self._disconnect()
                    await self._reconnect()
                except Exception as err:
                    self.error(f'Scheduled reconnect failed: {err}')

        self._companion_reconnect_task = asyncio.create_task(_reconnect_loop())

    def _stop_companion_reconnect(self) -> None:
        task = self._companion_reconnect_task
        self._companion_reconnect_task = None
        if task and not task.done():
            if asyncio.current_task() is not task:
                task.cancel()

    # ------------------------------------------------------------------
    # Capabilities
    # ------------------------------------------------------------------

    def _register_capabilities(self) -> None:
        self.register_capability_listener('onoff', self._on_onoff)
        self.register_capability_listener('speaker_next', self._on_speaker_next)
        self.register_capability_listener('speaker_prev', self._on_speaker_prev)
        self.register_capability_listener('speaker_playing', self._on_speaker_playing)
        self.register_capability_listener('volume_up', self._on_volume_up)
        self.register_capability_listener('volume_down', self._on_volume_down)
        self.register_capability_listener('volume_mute', self._on_volume_mute)
        self.register_capability_listener('volume_set', self._on_volume_set)
        self.register_capability_listener('remote_up', self._on_remote_up)
        self.register_capability_listener('remote_down', self._on_remote_down)
        self.register_capability_listener('remote_left', self._on_remote_left)
        self.register_capability_listener('remote_right', self._on_remote_right)
        self.register_capability_listener('remote_select', self._on_remote_select)
        self.register_capability_listener('remote_home', self._on_remote_home)
        self.register_capability_listener('remote_back', self._on_remote_back)
        self.register_capability_listener('remote_playpause', self._on_remote_playpause)
        self.register_capability_listener('button.restart', self._on_restart)

    async def _on_onoff(self, value: bool, *_) -> None:
        if self._atv is None:
            return
        if value:
            await self._atv.power.turn_on()
        else:
            await self._atv.power.turn_off()

    async def _on_speaker_next(self, *_) -> None:
        if self._atv is None:
            return
        await self._atv.remote_control.next()

    async def _on_speaker_prev(self, *_) -> None:
        if self._atv is None:
            return
        await self._atv.remote_control.previous()

    async def _on_speaker_playing(self, play: bool, *_) -> None:
        if self._atv is None:
            return
        # Use play_pause() for both — rc.play() can trigger Apple Music
        # instead of resuming the current media on newer Apple TVs
        await self._atv.remote_control.play_pause()

    async def _on_volume_up(self, *_) -> None:
        if self._atv is None:
            return
        try:
            await self._atv.audio.volume_up()
        except Exception as err:
            self.error(f'Volume up failed: {err}')

    async def _on_volume_down(self, *_) -> None:
        if self._atv is None:
            return
        try:
            await self._atv.audio.volume_down()
        except Exception as err:
            self.error(f'Volume down failed: {err}')

    async def _on_volume_mute(self, muted, *_) -> None:
        if self._atv is None:
            return
        try:
            if muted:
                await self._atv.audio.set_volume(0)
            self.log(f'Volume mute: {muted}')
        except Exception as err:
            self.error(f'Volume mute failed: {err}')

    async def _on_volume_set(self, volume: float, *_) -> None:
        if self._atv is None:
            return
        await self._atv.audio.set_volume(volume * 100)

    async def _on_remote_cmd(self, name, coro):
        if self._atv is None:
            return
        try:
            await coro
        except Exception as err:
            self.error(f'Remote {name} failed: {err}')

    async def _on_remote_up(self, *_): await self._on_remote_cmd('up', self._atv.remote_control.up())
    async def _on_remote_down(self, *_): await self._on_remote_cmd('down', self._atv.remote_control.down())
    async def _on_remote_left(self, *_): await self._on_remote_cmd('left', self._atv.remote_control.left())
    async def _on_remote_right(self, *_): await self._on_remote_cmd('right', self._atv.remote_control.right())
    async def _on_remote_select(self, *_): await self._on_remote_cmd('select', self._atv.remote_control.select())
    async def _on_remote_home(self, *_): await self._on_remote_cmd('home', self._atv.remote_control.home())
    async def _on_remote_back(self, *_): await self._on_remote_cmd('back', self._atv.remote_control.menu())
    async def _on_remote_playpause(self, *_): await self._on_remote_cmd('playpause', self._atv.remote_control.play_pause())

    async def _on_restart(self, *_) -> None:
        try:
            await self._disconnect()
            if self._airplay_logic:
                await self._airplay_logic.clear_now_playing()
            await self._reconnect()
        except Exception as err:
            self.error(err)


    # ------------------------------------------------------------------
    # Flow trigger hooks
    # ------------------------------------------------------------------

    async def trigger_now_playing_app_changed(self, bundle_id: str, display_name: str) -> None:
        app = self.homey.app
        if hasattr(app, 'apple_tv_flow') and app.apple_tv_flow:
            await app.apple_tv_flow.trigger_now_playing_app_changes(self, bundle_id, display_name)

    async def trigger_artwork_url_updated(self, image) -> None:
        app = self.homey.app
        if hasattr(app, 'apple_tv_flow') and app.apple_tv_flow:
            local_url = getattr(image, 'local_url', '') or ''
            cloud_url = getattr(image, 'cloud_url', '') or ''
            await app.apple_tv_flow.trigger_artwork_url_updated(self, local_url, cloud_url)

    async def trigger_companion_link_failed(self) -> None:
        app = self.homey.app
        if hasattr(app, 'apple_tv_flow') and app.apple_tv_flow:
            await app.apple_tv_flow.trigger_companion_link_failed(self)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _get_credentials(self, key: str = 'credentials') -> str | None:
        """Return stored credentials string from the device store, or None."""
        store = self.get_store()
        if not store:
            return None
        creds = store.get(key)
        if isinstance(creds, dict):
            self.error(f'Stored {key} in old format — please re-pair.')
            return None
        return creds if isinstance(creds, str) else None

homey_export = AppleTVDevice
